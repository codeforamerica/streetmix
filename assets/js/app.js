/*
 * Streetmix
 *
 * Front-end (mostly) by Marcin Wichary, Code for America fellow in 2013.
 *
 */

//= require vendor/modernizr
//= require vendor/moment
//= require vendor/jquery-2.1.0.min
//= require vendor/jquery.cookie

//= require_self
//= require_tree constants
//= require_tree util
//= require misc
//= require_tree app
//= require_tree segments
//= require_tree streets
//= require_tree users

var app = {};
