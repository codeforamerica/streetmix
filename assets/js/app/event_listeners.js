function _addEventListeners () {
  document.querySelector('#street-section-left-building').addEventListener('pointerenter', _onBuildingMouseEnter)
  document.querySelector('#street-section-left-building').addEventListener('pointerleave', _onBuildingMouseLeave)
  document.querySelector('#street-section-right-building').addEventListener('pointerenter', _onBuildingMouseEnter)
  document.querySelector('#street-section-right-building').addEventListener('pointerleave', _onBuildingMouseLeave)

  document.querySelector('.info-bubble').addEventListener('pointerenter', _infoBubble.onMouseEnter)
  document.querySelector('.info-bubble').addEventListener('pointerleave', _infoBubble.onMouseLeave)
  document.querySelector('.info-bubble').addEventListener('pointerdown', _infoBubble.onTouchStart)

  document.querySelector('#no-connection-try-again').addEventListener('pointerdown', _nonblockingAjaxTryAgain)

  // The following do not seem to work on pointerdown
  // click should also be fired by other input methods
  document.querySelector('#new-street-default').addEventListener('click', _onNewStreetDefaultClick)
  document.querySelector('#new-street-empty').addEventListener('click', _onNewStreetEmptyClick)
  document.querySelector('#new-street-last').addEventListener('click', _onNewStreetLastClick)

  window.addEventListener('storage', _onStorageChange)

  document.querySelector('#sign-out-link').addEventListener('pointerdown', _onSignOutClick)

  if (system.pageVisibility) {
    document.addEventListener(system.visibilityChange, _onVisibilityChange, false)
  } else {
    window.addEventListener('focus', _onWindowFocus)
    window.addEventListener('blur', _onWindowBlur)
  }

  window.addEventListener('beforeunload', _onWindowBeforeUnload)

  document.querySelector('#undo').addEventListener('pointerdown', _undo)
  document.querySelector('#redo').addEventListener('pointerdown', _redo)

  if (!app.readOnly) {
    document.querySelector('#street-width-read').addEventListener('pointerdown', _onStreetWidthClick)
    document.querySelector('#street-width').addEventListener('change', _onStreetWidthChange)
  }

  window.addEventListener('resize', _onResize)

  // This listener hides the info bubble when the mouse leaves the
  // document area. Do not normalize it to a pointerleave event
  // because it doesn't make sense for other pointer types
  document.addEventListener('mouseleave', _onBodyMouseOut)

  window.addEventListener('pointerdown', _onBodyMouseDown)
  window.addEventListener('pointermove', _onBodyMouseMove)
  window.addEventListener('pointerup', _onBodyMouseUp)
  window.addEventListener('keydown', _onGlobalKeyDown)

}
