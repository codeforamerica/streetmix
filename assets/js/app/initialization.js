var initializing = false
var bodyLoaded
var readyStateCompleteLoaded

var abortEverything

var TRACK_ACTION_TOUCH_CAPABLE = 'Touch capability detected'

// Some things are placed on the generic Stmx app object to keep it out of global scope
// Do this as little as possible. Eventually, code becomes a collection of
// individual modules that are require()'d by browserify
var Stmx = {}

Stmx.preInit = function () {
  initializing = true
  ignoreStreetChanges = true

  var language = window.navigator.userLanguage || window.navigator.language
  if (language) {
    language = language.substr(0, 2).toUpperCase()
    _updateSettingsFromCountryCode(language)
  }
}

Stmx.init = function () {
  /* global Locale */
  if (!debug.forceUnsupportedBrowser) {
    // TODO temporary ban
    if ((navigator.userAgent.indexOf('Opera') != -1) ||
      (navigator.userAgent.indexOf('Internet Explorer') != -1) ||
      (navigator.userAgent.indexOf('MSIE') != -1)) {
      mode = MODES.UNSUPPORTED_BROWSER
      _processMode()
      return
    }
  }

  _initGallery() // formerly _fillDom()
  _fillEmptySegments()
  _prepareSegmentInfo()

  // TODO make it better
  // Related to Enter to 404 bug in Chrome
  $.ajaxSetup({ cache: false })

  readyStateCompleteLoaded = false
  document.addEventListener('readystatechange', _onReadyStateChange)

  bodyLoaded = false
  window.addEventListener('load', _onBodyLoad)

  _processUrl()
  _processMode()

  if (abortEverything) {
    return
  }

  // Asynchronously loading…

  // …detecting country from IP for units and left/right-hand driving
  if ((mode == MODES.NEW_STREET) || (mode == MODES.NEW_STREET_COPY_LAST)) {
    _detectGeolocation()
  } else {
    geolocationLoaded = true
  }

  // …sign in info from our API (if not previously cached) – and subsequent
  // street data if necessary (depending on the mode)
  _loadSignIn()

// Note that we are waiting for sign in and image info to show the page,
// but we give up on country info if it’s more than 1000ms.
}

function _onEverythingLoaded () {
  switch (mode) {
    case MODES.NEW_STREET_COPY_LAST:
      _onNewStreetLastClick()
      break
  }

  _showWelcome()

  _onResize()
  _resizeStreetWidth()
  _updateStreetName()
  _createDomFromData()
  _segmentsChanged()

  initializing = false
  ignoreStreetChanges = false
  lastStreet = _trimStreetData(street)

  _updatePageUrl()
  _buildStreetWidthMenu()
  _addScrollButtons(document.querySelector('#palette'))
  _addScrollButtons(document.querySelector('#gallery .streets'))
  _addEventListeners()

  var event = new CustomEvent('stmx:everything_loaded')
  window.dispatchEvent(event)

  if (mode == MODES.USER_GALLERY) {
    _showGallery(galleryUserId, true)
  } else if (mode == MODES.GLOBAL_GALLERY) {
    _showGallery(null, true)
  } else if (mode == MODES.ABOUT) {
    aboutDialog.show()
  }

  if (promoteStreet) {
    _remixStreet()
  }

  // Track touch capability in Google Analytics
  if (system.touch === true) {
    trackEvent('System', TRACK_ACTION_TOUCH_CAPABLE, null, null, true)
  }
}

function _onBodyLoad () {
  bodyLoaded = true

  document.querySelector('#loading-progress').value++
  _checkIfEverythingIsLoaded()
}

function _onReadyStateChange () {
  if (document.readyState == 'complete') {
    readyStateCompleteLoaded = true

    document.querySelector('#loading-progress').value++
    _checkIfEverythingIsLoaded()
  }
}
