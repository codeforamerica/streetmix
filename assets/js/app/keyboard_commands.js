var TRACK_LABEL_KEYBOARD = 'Keyboard'

var KEYS = {
  LEFT_ARROW: 37,
  RIGHT_ARROW: 39,
  ENTER: 13,
  BACKSPACE: 8,
  DELETE: 46,
  ESC: 27,
  Y: 89,
  Z: 90,
  EQUAL: 187, // = or +
  EQUAL_ALT: 61, // Firefox
  PLUS_KEYPAD: 107,
  MINUS: 189,
  MINUS_ALT: 173, // Firefox
  MINUS_KEYPAD: 109
}

function _onGlobalKeyDown (event) {
  if (_isFocusOnBody()) {
    _onBodyKeyDown(event)
  }

  switch (event.keyCode) {
    case KEYS.ESC:
      if (document.querySelector('#welcome').classList.contains('visible')) {
        _hideWelcome()
      } else if (isAnyDialogVisible()) {
        hideAllDialogs()
      } else if (draggingType == DRAGGING_TYPE_RESIZE) {
        _handleSegmentResizeCancel()
      } else if (draggingType == DRAGGING_TYPE_MOVE) {
        _handleSegmentMoveCancel()
      } else if (isAnyMenuVisible() === true) {
        hideAllMenus()
      } else if (document.querySelector('#status-message').classList.contains('visible')) {
        hideStatusMessage()
      } else if (_infoBubble.visible && _infoBubble.descriptionVisible) {
        _infoBubble.hideDescription()
      } else if (_infoBubble.visible) {
        _infoBubble.hide()
        _infoBubble.hideSegment(false)
      } else if (document.body.classList.contains('gallery-visible')) {
        _hideGallery(false)
      } else if (signedIn) {
        _showGallery(signInData.userId, false)
      } else {
        return
      }

      event.preventDefault()
      break
  }
}

function _onBodyKeyDown (event) {
  switch (event.keyCode) {
    case KEYS.EQUAL:
    case KEYS.EQUAL_ALT:
    case KEYS.PLUS_KEYPAD:
    case KEYS.MINUS:
    case KEYS.MINUS_ALT:
    case KEYS.MINUS_KEYPAD:
      if (event.metaKey || event.ctrlKey || event.altKey) {
        return
      }

      var negative = (event.keyCode == KEYS.MINUS) ||
        (event.keyCode == KEYS.MINUS_ALT) ||
        (event.keyCode == KEYS.MINUS_KEYPAD)

      var hoveredEl = _getHoveredEl()
      if (hoveredEl) {
        if (hoveredEl.classList.contains('segment')) {
          _incrementSegmentWidth(hoveredEl, !negative, event.shiftKey)
        } else if (hoveredEl.id == 'street-section-left-building') {
          _changeBuildingHeight(true, !negative)
        } else if (hoveredEl.id == 'street-section-right-building') {
          _changeBuildingHeight(false, !negative)
        }
        event.preventDefault()

        trackEvent('Interaction', TRACK_ACTION_CHANGE_WIDTH,
          TRACK_LABEL_KEYBOARD, null, true)
      }
      break
    case KEYS.BACKSPACE:
    case KEYS.DELETE:
      if (event.metaKey || event.ctrlKey || event.altKey) {
        return
      }

      var segmentHoveredEl = _getHoveredSegmentEl()
      _removeSegment(segmentHoveredEl, event.shiftKey)

      trackEvent('Interaction', TRACK_ACTION_REMOVE_SEGMENT,
        TRACK_LABEL_KEYBOARD, null, true)

      event.preventDefault()
      break
    case KEYS.LEFT_ARROW:
      if (event.metaKey || event.ctrlKey || event.altKey) {
        return
      }
      _scrollStreet(true, event.shiftKey)
      event.preventDefault()
      break
    case KEYS.RIGHT_ARROW:
      if (event.metaKey || event.ctrlKey || event.altKey) {
        return
      }
      _scrollStreet(false, event.shiftKey)
      event.preventDefault()
      break
    case KEYS.Z:
      if (!event.shiftKey && (event.metaKey || event.ctrlKey)) {
        _undo()
        event.preventDefault()
      } else if (event.shiftKey && (event.metaKey || event.ctrlKey)) {
        _redo()
        event.preventDefault()
      }
      break
    case KEYS.Y:
      if (event.metaKey || event.ctrlKey) {
        _redo()
        event.preventDefault()
      }
      break
  }
}

function _getHoveredSegmentEl () {
  var el = document.querySelector('.segment.hover')
  return el
}

function _getHoveredEl () {
  var el = document.querySelector('.hover')
  return el
}
