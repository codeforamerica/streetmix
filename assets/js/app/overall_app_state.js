var abortEverything;

var bodyLoaded;
var readyStateCompleteLoaded;
var serverContacted;
