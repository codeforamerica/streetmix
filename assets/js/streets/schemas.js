var LATEST_SCHEMA_VERSION = 16;
// 1: starting point
// 2: adding leftBuildingHeight and rightBuildingHeight
// 3: adding leftBuildingVariant and rightBuildingVariant
// 4: adding transit shelter elevation
// 5: adding another lamp type (traditional)
// 6: colored streetcar lanes
// 7: colored bus and light rail lanes
// 8: colored bike lane
// 9: second car type: truck
// 10: sidewalk density
// 11: unify median and planting strip into divider
// 12: getting rid of small tree
// 13: bike rack elevation
// 14: wayfinding has three types
// 15: sidewalks have rand seed
// 16: stop saving undo stack

function _incrementSchemaVersion(street) {
  if (!street.schemaVersion) {
    street.schemaVersion = 1;
  }

  switch (street.schemaVersion) {
    case 1:
      street.leftBuildingHeight = DEFAULT_BUILDING_HEIGHT_LEFT;
      street.rightBuildingHeight = DEFAULT_BUILDING_HEIGHT_RIGHT;
      break;
    case 2:
      street.leftBuildingVariant = DEFAULT_BUILDING_VARIANT_LEFT;
      street.rightBuildingVariant = DEFAULT_BUILDING_VARIANT_RIGHT;
      break;
    case 3:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'transit-shelter') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['transit-shelter-elevation'] = 'street-level';
          segment.variantString = _getVariantString(variant);
        }
      }
      break;
    case 4:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'sidewalk-lamp') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['lamp-type'] = 'modern';
          segment.variantString = _getVariantString(variant);
        }
      }
      break;
    case 5:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'streetcar') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['public-transit-asphalt'] = 'regular';
          segment.variantString = _getVariantString(variant);
        }
      }
      break;
    case 6:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if ((segment.type == 'bus-lane') || (segment.type == 'light-rail')) {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['public-transit-asphalt'] = 'regular';
          segment.variantString = _getVariantString(variant);
        }
      }
      break;
    case 7:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'bike-lane') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['bike-asphalt'] = 'regular';
          segment.variantString = _getVariantString(variant);
        }
      }
      break;
    case 8:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'drive-lane') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['car-type'] = 'car';
          segment.variantString = _getVariantString(variant);
        }
      }
      break;
    case 9:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'sidewalk') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['sidewalk-density'] = 'normal';
          segment.variantString = _getVariantString(variant);
        }
      }
      break;
    case 10:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'planting-strip') {
          segment.type = 'divider';

          if (segment.variantString == '') {
            segment.variantString = 'planting-strip';
          };
        } else if (segment.type == 'small-median') {
          segment.type = 'divider';
          segment.variantString = 'median';
        }
      }
      break;
    case 11:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'divider') {
          if (segment.variantString == 'small-tree') {
            segment.variantString = 'big-tree';
          };
        } else if (segment.type == 'sidewalk-tree') {
          if (segment.variantString == 'small') {
            segment.variantString = 'big';
          };
        }
      }
      break;
    case 12:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'sidewalk-bike-rack') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['bike-rack-elevation'] = 'sidewalk';
          segment.variantString =  _getVariantString(variant);
        }
      }
      break;
    case 13:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'sidewalk-wayfinding') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['wayfinding-type'] = 'large';
          segment.variantString =  _getVariantString(variant);
        }
      }
      break;
    case 14:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'sidewalk') {
          segment.randSeed = 35;
        }
      }
      break;
    case 15:
      undoStack = [];
      undoPosition = 0;
      break;
  }

  street.schemaVersion++;
}

function _updateToLatestSchemaVersion(street) {
  var updated = false;
  while (!street.schemaVersion || (street.schemaVersion < LATEST_SCHEMA_VERSION)) {
    _incrementSchemaVersion(street);
    updated = true;
  }

  return updated;
}
