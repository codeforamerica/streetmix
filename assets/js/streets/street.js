var street = {
  schemaVersion: LATEST_SCHEMA_VERSION,

  id: null,
  creatorId: null,
  namespacedId: null,
  originalStreetId: null, // id of the street the current street is remixed from (could be null)
  name: null,
  editCount: null, // Since new street or remix · FIXME: can be null (meaning = don’t touch), but this will change

  width: null,
  occupiedWidth: null, // Can be recreated, do not save
  remainingWidth: null, // Can be recreated, do not save

  leftBuildingHeight: null,
  rightBuildingHeight: null,
  leftBuildingVariant: null,
  rightBuildingVariant: null,

  segments: [],

  units: null
};
