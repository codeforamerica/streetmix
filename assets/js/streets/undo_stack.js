var UNDO_LIMIT = 1000

var FLAG_SAVE_UNDO = false; // true to save undo with street data, false to not save undo

var undoStack = []
var undoPosition = 0
var ignoreStreetChanges = false

function _undoRedo (undo) {
  if (undo && !_isUndoAvailable()) {
    showStatusMessage(msg('STATUS_NOTHING_TO_UNDO'))
  } else if (!undo && !_isRedoAvailable()) {
    showStatusMessage(msg('STATUS_NOTHING_TO_REDO'))
  } else {
    if (undo) {
      undoStack[undoPosition] = _trimStreetData(street)
      undoPosition--
    } else {
      undoPosition++
    }
    street = _.cloneDeep(undoStack[undoPosition])
    _setUpdateTimeToNow()

    _infoBubble.hide()
    _infoBubble.hideSegment()
    _infoBubble.dontConsiderShowing()

    _updateEverything(true)
    hideStatusMessage()
  }

  trackEvent('INTERACTION', 'UNDO', null, null, true)
}

function _clearUndoStack () {
  undoStack = []
  undoPosition = 0
  _updateUndoButtons()
}

function _undo () {
  _undoRedo(true)
}

function _redo () {
  _undoRedo(false)
}

function _trimUndoStack () {
  // TODO optimize
  while (undoPosition >= UNDO_LIMIT) {
    undoPosition--
    undoStack = undoStack.slice(1)
  }
}

function _createNewUndo () {
  // This removes future undo path in case we undo a few times and then do
  // something undoable.
  undoStack = undoStack.splice(0, undoPosition)
  undoStack[undoPosition] = _.cloneDeep(lastStreet)
  undoPosition++

  _trimUndoStack()
  _unifyUndoStack()
}

function _createNewUndoIfNecessary (lastStreet, currentStreet) {
  if (lastStreet.name != currentStreet.name) {
    return
  }

  _createNewUndo()
}

function _isUndoAvailable () {
  // Don’t allow undo/redo unless you own the street

  return (undoPosition > 0) && !remixOnFirstEdit
}

function _isRedoAvailable () {
  // Don’t allow undo/redo unless you own the street

  return (undoPosition < undoStack.length - 1) && !remixOnFirstEdit
}

function _updateUndoButtons () {
  document.querySelector('#undo').disabled = !_isUndoAvailable()
  document.querySelector('#redo').disabled = !_isRedoAvailable()
}

function _unifyUndoStack () {
  for (var i = 0; i < undoStack.length; i++) {
    undoStack[i].id = street.id
    undoStack[i].name = street.name
    undoStack[i].namespacedId = street.namespacedId
    undoStack[i].creatorId = street.creatorId
    undoStack[i].updatedAt = street.updatedAt
  }
}
