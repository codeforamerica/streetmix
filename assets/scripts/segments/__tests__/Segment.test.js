/* eslint-env jest */
import React from 'react'
import { fireEvent, getByTestId, cleanup } from '@testing-library/react'
import { renderWithRedux } from '../../../../test/helpers/render'
import Segment from '../Segment'
import { getSpriteDef, getSegmentInfo, getSegmentVariantInfo } from '../info'
import SEGMENT_INFO from '../info.json'
import { KEYS } from '../../app/keys'
import { infoBubble } from '../../info_bubble/info_bubble'
import { setLastStreet } from '../../streets/data_model'

// Replace all increment resolution with a simple value of 1
const __TEST_RESIZE_INCREMENT = 1

jest.mock('react-dnd')
jest.mock('../info')
jest.mock('../view', () => {
  const actual = jest.requireActual('../view')
  return {
    ...actual,
    getVariantInfoDimensions: jest.fn(() => ({ left: 0, right: 100, center: 50 }))
  }
})
jest.mock('../resizing', () => {
  const actual = jest.requireActual('../resizing')
  return {
    ...actual,
    // Note: __TEST_RESIZE_INCREMENT cannot be used here because jest wants to
    // hoist this function to before it is declared, so just hard-code 1
    resolutionForResizeType: jest.fn(() => 1)
  }
})
jest.mock('../../app/load_resources')
jest.mock('../../app/routing', () => {})
jest.mock('../../info_bubble/info_bubble')

// ToDo: Remove this once refactoring of redux action saveStreetToServerIfNecessary is complete
jest.mock('../../streets/data_model', () => {
  const actual = jest.requireActual('../../streets/data_model')
  return {
    ...actual,
    saveStreetToServerIfNecessary: jest.fn()
  }
})

describe('Segment', () => {
  let variantString, type, variant, currentWidth, increment, activeElement, segment

  beforeEach(() => {
    variantString = 'inbound|regular'
    type = 'streetcar'
    variant = SEGMENT_INFO[type].details[variantString]
    currentWidth = 200
    increment = __TEST_RESIZE_INCREMENT
    activeElement = 0
    segment = { type, variantString, segmentType: type, id: '1', width: currentWidth, randSeed: 1 }
    const segmentInfo = { name: 'Segment', nameKey: 'key', zIndex: 1 }
    getSegmentInfo.mockImplementation(() => segmentInfo)
    getSegmentVariantInfo.mockImplementation(() => variant)
    getSpriteDef.mockImplementation(() => ({ id: 'markings--straight-inbound', width: 4, offsetY: 11.12 }))
  })

  afterEach(cleanup)

  it('renders correctly', () => {
    const wrapper = renderWithRedux(<Segment segment={segment} actualWidth={currentWidth} dataNo={activeElement} updateSegmentData={jest.fn()} connectDragPreview={jest.fn()} />, { initialState: { ui: { activeSegment: activeElement }, street: { segments: [segment] } } })
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('shows the info bubble on mouseover', () => {
    const wrapper = renderWithRedux(<Segment segment={segment} actualWidth={currentWidth} dataNo={activeElement} updateSegmentData={jest.fn()} connectDragPreview={jest.fn()} />, { initialState: { ui: { activeSegment: activeElement }, street: { segments: [segment] } } })
    fireEvent.mouseOver(getByTestId(wrapper.container, 'segment'))
    expect(infoBubble.considerShowing).toHaveBeenCalledTimes(1)
  })

  it('hides the info bubble on mouseleave', () => {
    const wrapper = renderWithRedux(<Segment segment={segment} actualWidth={currentWidth} dataNo={activeElement} updateSegmentData={jest.fn()} connectDragPreview={jest.fn()} />, { initialState: { ui: { activeSegment: activeElement }, street: { segments: [segment] } } })
    fireEvent.mouseOver(getByTestId(wrapper.container, 'segment'))
    fireEvent.mouseLeave(getByTestId(wrapper.container, 'segment'))
    expect(infoBubble.dontConsiderShowing).toHaveBeenCalledTimes(1)
  })

  describe('keyboard events', () => {
    it('KEY.MINUS decreases the width of the segment', () => {
      const wrapper = renderWithRedux(<Segment segment={segment} actualWidth={currentWidth} dataNo={activeElement} updateSegmentData={jest.fn()} connectDragPreview={jest.fn()} />, { initialState: { ui: { activeSegment: activeElement }, street: { segments: [segment] } } })
      fireEvent.mouseOver(getByTestId(wrapper.container, 'segment'))
      fireEvent.keyDown(document, { key: 'Minus', keyCode: KEYS.MINUS, code: KEYS.MINUS, charCode: KEYS.MINUS })
      expect(wrapper.store.getState().street.segments[activeElement].width).toEqual(currentWidth - increment)
    })

    it('KEY.EQUAL increases the width of the segment', () => {
      const wrapper = renderWithRedux(<Segment segment={segment} actualWidth={currentWidth} dataNo={activeElement} updateSegmentData={jest.fn()} connectDragPreview={jest.fn()} />, { initialState: { ui: { activeSegment: activeElement }, street: { segments: [segment] } } })
      fireEvent.mouseOver(getByTestId(wrapper.container, 'segment'))
      fireEvent.keyDown(document, { key: 'Equal', keyCode: KEYS.EQUAL, code: KEYS.EQUAL, charCode: KEYS.EQUAL })
      expect(wrapper.store.getState().street.segments[activeElement].width).toEqual(currentWidth + increment)
    })

    it('removes segment when delete key is pressed', () => {
      const wrapper = renderWithRedux(<Segment segment={segment} actualWidth={currentWidth} dataNo={activeElement} updateSegmentData={jest.fn()} connectDragPreview={jest.fn()} />, { initialState: { ui: { activeSegment: activeElement }, street: { segments: [segment] } } })
      setLastStreet() // ToDo: needs to be refactored
      fireEvent.mouseOver(getByTestId(wrapper.container, 'segment'))
      fireEvent.keyDown(document, { key: 'Delete', keyCode: KEYS.DELETE, code: KEYS.DELETE, charCode: KEYS.DELETE })
      expect(infoBubble.hide).toHaveBeenCalledTimes(1)
      expect(infoBubble.hideSegment).toHaveBeenCalledTimes(1)
      expect(wrapper.store.getState().street.segments.length).toEqual(0)
    })

    it('removes all segments when shift+delete keys are pressed', () => {
      const wrapper = renderWithRedux(<Segment segment={segment} actualWidth={currentWidth} dataNo={activeElement} updateSegmentData={jest.fn()} connectDragPreview={jest.fn()} />, { initialState: { ui: { activeSegment: activeElement }, street: { segments: [segment] } } })
      setLastStreet() // ToDo: needs to be refactored
      fireEvent.mouseOver(getByTestId(wrapper.container, 'segment'))
      fireEvent.keyDown(document, { key: 'Delete', keyCode: KEYS.DELETE, code: KEYS.DELETE, charCode: KEYS.DELETE, shiftKey: true })
      expect(infoBubble.hide).toHaveBeenCalledTimes(2) // toDo: should this be 1?
      expect(infoBubble.hideSegment).toHaveBeenCalledTimes(1)
      expect(wrapper.store.getState().street.segments.length).toEqual(0)
    })
  })
})
