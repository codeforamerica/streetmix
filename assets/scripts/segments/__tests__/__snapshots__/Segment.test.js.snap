// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Segment renders correctly 1`] = `
<DocumentFragment>
  <div
    class="segment hover show-drag-handles"
    data-testid="segment"
    style="width: 2400px; z-index: 1; transform: translateX(undefinedpx);"
  >
    <div
      class="segment-label-container"
    >
      <span
        class="segment-label"
      >
        Segment
      </span>
      <span
        class="segment-width"
      >
        <span>
          60 m
        </span>
      </span>
      <span
        class="segment-grid units-metric"
      />
    </div>
    <span
      class="drag-handle drag-handle-left"
    >
      <svg
        class="svg-inline--fa fa-chevron-left"
      />
    </span>
    <span
      class="drag-handle drag-handle-right"
    >
      <svg
        class="svg-inline--fa fa-chevron-right"
      />
    </span>
    <div
      class="segment-canvas-container"
    >
      <canvas
        class="segment-image"
        height="445"
        style="width: 2400px; height: 445px; left: 0px;"
        width="2400"
      />
    </div>
    <div
      class="hover-bk"
    />
  </div>
</DocumentFragment>
`;
